code/modules/vehicles/tank.dm

```diff
@@ -0,0 +1,491 @@
+	var/secondary_weapon_type = /obj/item/tank_weapon/secondary_weapon
+	var/primary_weapon_dir = null //So that the guns swivel independantly
+	var/secondary_weapon_dir = null
+	var/atom/firing_target = null //Shooting code, at whom are we firing?
+	var/firing_primary_weapon = FALSE
+	var/firing_secondary_weapon = FALSE
+	var/last_drive_sound = 0 //Engine noises.
+	var/list/passengers = list() //People who are in the tank without gunning / driving. This allows for things like jeeps and APCs in future
+	var/max_passengers = 5 //This seems sane to me, change if you don't agree.
+	var/lights_on = FALSE //Tanks start with lights off
+
+
+/obj/vehicle/tank/tiny //SQUEEEE
+	name = "Mk-4 'pint pot' tank"
+	desc = "An adorable chunk of metal with an alarming amount of firepower designed to crush, immolate, destroy and maim anything that nanotrasen wants it to. This model contains advanced bluespace technology which allows a tardis like amount of room on the inside"
+	icon = 'icons/obj/tinytank.dmi'
+	turret_icon = 'icons/obj/tinytank_gun.dmi'
+	icon_state = "tank"
+	pixel_x = -16 //Stops marines from treading on it...d'aww
+	pixel_y = -8
+	max_passengers = 2 //Bluespace's one hell of a drug.
+
+/obj/vehicle/tank/apc //SQUEEEE
+	name = "M557 Armoured Personnel Carrier"
+	desc = "A heavily armoured vehicle with light armaments. Though it doesn't pack the punch of a tank its utility launcher can provide cover for troops and allow for rapid deployment with minimal casualties."
+	icon = 'icons/obj/tinytank.dmi'
+	turret_icon = 'icons/obj/tinytank_gun.dmi'
+	turret_icon_state = "apc_turret"
+	icon_state = "apc"
+	move_delay = 0.5 SECONDS //Faster because it's lightweight
+	pixel_x = -16
+	pixel_y = -8
+	max_passengers = 4 //It's a big carrier designed to get marines into the operation safely.
+	primary_weapon_type = /obj/item/tank_weapon/apc_cannon //Only has a utility launcher, no offense as standard.
+	secondary_weapon_type = null
+
+/obj/vehicle/tank/examine(mob/user)
+	. = ..()
+	to_chat(user, "<b><span class='notice>To fire its main cannon, <i>ctrl</i> click a tile</b></span>")
+	to_chat(user, "<b><span class='notice>To fire its secondary weapon, click a tile</b></span>")
+	to_chat(user, "<i><span class='notice>It's currently holding [passengers.len] / [max_passengers] passengers</i></span>")
+
+/obj/turret_overlay
+	name = "Tank gun turret"
+	desc = "The shooty bit on a tank."
+	icon = 'icons/obj/tank_gun.dmi'
+	icon_state = "turret"
+	layer = ABOVE_MOB_LAYER+0.1
+	animate_movement = TRUE //So it doesnt just ping back and forth and look all stupid
+	mouse_opacity = FALSE //It's an overlay
+
+/obj/vehicle/tank/Initialize()
+	. = ..()
+	turret_overlay = new()
+	turret_overlay.icon = turret_icon
+	turret_overlay.icon_state = turret_icon_state
+	update_icon()
+	if(primary_weapon_type)
+		primary_weapon = new primary_weapon_type(src) //Make our guns
+	if(secondary_weapon_type)
+		secondary_weapon = new secondary_weapon_type(src)
+	primary_weapon?.owner = src
```

Comment:
Put this inside the `if(primary_weapon_type)`
A `null = src` operation would be bad.
Same for the secondary one.